compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -433,22 +428,19 @@ impl<'a, 'b, 'tcx> TypeFolder<'tcx> for AssocTypeNormalizer<'a, 'b, 'tcx> {
                 normalized_ty
             }
 
-            ty::Projection(data) if !data.trait_ref(self.tcx()).has_escaping_bound_vars() => {
-                // Okay, so you thought the previous branch was hacky. Well, to
-                // extend upon this, when the *trait ref* doesn't have escaping
-                // bound vars, but the associated item *does* (can only occur
-                // with GATs), then we might still be able to project the type.
-                // For this, we temporarily replace the bound vars with
-                // placeholders. Note though, that in the case that we still
-                // can't project for whatever reason (e.g. self type isn't
-                // known enough), we *can't* register an obligation and return
-                // an inference variable (since then that obligation would have
-                // bound vars and that's a can of worms). Instead, we just
-                // give up and fall back to pretending like we never tried!
+            ty::Projection(data) => {
+                // If there are escaping bound vars, we temporarily replace the
+                // bound vars with placeholders. Note though, that in the cas
+                // that we still can't project for whatever reason (e.g. self
+                // type isn't known enough), we *can't* register an obligation
+                // and return an inference variable (since then that obligation
+                // would have bound vars and that's a can of worms). Instead,
+                // we just give up and fall back to pretending like we never tried!
```

Comment:
I remember this being discussed on Zulip - is this intended to be a 'temporary' hack until we figure out an approach for registering obligations with bound bars?

Reply:
Yes and no. This specific behavior of giving up if we can't normalize is, yes, somewhat of a hack. Trying to tread the line between normalizing *more* and not blowing things up.

That said, it's not clear if the "correct" approach is registering obligations with bound vars, or maybe just a lazy normalization approach.

Reply:
Could you update the comment to note that this is not necessarily the final approach to dealing with obligations?